internal/controller/changetransferpolicy_controller.go

```diff
@@ -155,6 +162,227 @@ func (r *ChangeTransferPolicyReconciler) Reconcile(ctx context.Context, req ctrl
+		return promoterv1alpha1.History{}, false, fmt.Errorf("failed to get trailers for SHA %q: %w", sha, err)
+	}
+
+	// Skip no-op commits
+	if activeTrailers[constants.TrailerNoOp] == "true" {
+		return promoterv1alpha1.History{}, false, nil
+	}
+
+	historyEntry := promoterv1alpha1.History{
+		Proposed:    promoterv1alpha1.CommitBranchStateHistoryProposed{},
+		Active:      promoterv1alpha1.CommitBranchState{},
+		PullRequest: &promoterv1alpha1.PullRequestCommonStatus{},
+	}
+
+	r.populateActiveMetadata(ctx, &historyEntry, sha, gitOperations)
+	r.populateProposedMetadata(ctx, &historyEntry, activeTrailers, gitOperations)
+	r.populatePullRequestMetadata(ctx, &historyEntry, activeTrailers)
+	r.populateCommitStatuses(ctx, &historyEntry, activeTrailers)
+
+	return historyEntry, true, nil
+}
+
+// populateActiveMetadata populates the active metadata for a history entry
+func (r *ChangeTransferPolicyReconciler) populateActiveMetadata(ctx context.Context, h *promoterv1alpha1.History, sha string, gitOperations *git.EnvironmentOperations) {
+	logger := log.FromContext(ctx)
+
+	if activeHydrated, err := gitOperations.GetShaMetadataFromGit(ctx, sha); err != nil {
+		logger.Error(err, "failed to get hydrated commit metadata for active SHA", "activeSha", sha)
+	} else {
+		h.Active.Hydrated = activeHydrated
+		h.Active.Hydrated.Body = removeKnownTrailers(h.Active.Hydrated.Body)
+	}
+
+	if activeDry, err := gitOperations.GetShaMetadataFromFile(ctx, sha); err != nil {
+		logger.Error(err, "failed to get dry commit metadata for active SHA", "activeSha", sha)
+	} else {
+		h.Active.Dry = activeDry
+	}
+}
+
+// populateProposedMetadata populates the proposed metadata for a history entry
+func (r *ChangeTransferPolicyReconciler) populateProposedMetadata(ctx context.Context, h *promoterv1alpha1.History, activeTrailers map[string]string, gitOperations *git.EnvironmentOperations) {
+	logger := log.FromContext(ctx)
+
+	proposedHydratedSha := activeTrailers[constants.TrailerShaHydratedProposed]
+	if proposedHydratedSha == "" {
+		logger.V(4).Info("No Sha-Hydrated-Proposed trailer found")
+		return
+	}
+
+	if meta, err := gitOperations.GetShaMetadataFromGit(ctx, proposedHydratedSha); err != nil {
+		logger.Error(err, "failed to get hydrated commit metadata for proposed SHA", "proposedSha", proposedHydratedSha)
+	} else {
+		h.Proposed.Hydrated = meta
+	}
+}
+
+// populatePullRequestMetadata populates the pull request metadata for a history entry
+func (r *ChangeTransferPolicyReconciler) populatePullRequestMetadata(ctx context.Context, h *promoterv1alpha1.History, activeTrailers map[string]string) {
+	logger := log.FromContext(ctx)
+
+	if pullRequestID := activeTrailers[constants.TrailerPullRequestID]; pullRequestID != "" {
+		h.PullRequest.ID = pullRequestID
+	} else {
+		logger.V(4).Info("No PullRequest-ID found in trailers")
+	}
+
+	if pullRequestUrl := activeTrailers[constants.TrailerPullRequestUrl]; pullRequestUrl != "" {
+		h.PullRequest.Url = pullRequestUrl
```

Comment:
We should probably consider this value trusted, but as an extra layer of protection maybe we should validate that the URL has either an http or https schema.

Reply:
Yeah, I think log it and leave it empty